ssl/ssl_quic.c

```diff
@@ -0,0 +1,250 @@
+{
+    return ssl->quic_read_level;
+}
+
+OSSL_ENCRYPTION_LEVEL SSL_quic_write_level(const SSL *ssl)
+{
+    return ssl->quic_write_level;
+}
+
+int SSL_provide_quic_data(SSL *ssl, OSSL_ENCRYPTION_LEVEL level,
+                          const uint8_t *data, size_t len)
+{
+    size_t l;
+
+    if (!SSL_IS_QUIC(ssl)) {
+        SSLerr(SSL_F_SSL_PROVIDE_QUIC_DATA, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+
+    /* Level can be different than the current read, but not less */
+    if (level < ssl->quic_read_level
+            || (ssl->quic_input_data_tail != NULL && level < ssl->quic_input_data_tail->level)) {
+        SSLerr(SSL_F_SSL_PROVIDE_QUIC_DATA, SSL_R_WRONG_ENCRYPTION_LEVEL_RECEIVED);
+        return 0;
+    }
+
+    /* Split the QUIC messages up, if necessary */
+    while (len > 0) {
+        QUIC_DATA *qd;
+        const uint8_t *p = data + 1;
+
+        n2l3(p, l);
+        l += SSL3_HM_HEADER_LENGTH;
+
+        if (l > len) {
+            SSLerr(SSL_F_SSL_PROVIDE_QUIC_DATA, SSL_R_BAD_DATA_LENGTH);
+            return 0;
+        }
+
+        qd = OPENSSL_malloc(sizeof(QUIC_DATA) + l);
+        if (qd == NULL) {
+            SSLerr(SSL_F_SSL_PROVIDE_QUIC_DATA, SSL_R_INTERNAL_ERROR);
+            return 0;
+        }
+
+        qd->next = NULL;
+        qd->length = l;
+        qd->level = level;
+        memcpy((void*)(qd + 1), data, l);
+        if (ssl->quic_input_data_tail != NULL)
+            ssl->quic_input_data_tail->next = qd;
+        else
+            ssl->quic_input_data_head = qd;
+        ssl->quic_input_data_tail = qd;
+
+        data += l;
+        len -= l;
+    }
+
+    return 1;
+}
+
+int SSL_CTX_set_quic_method(SSL_CTX *ctx, const SSL_QUIC_METHOD *quic_method)
+{
+    switch (ctx->method->version) {
+    case DTLS1_VERSION:
+    case DTLS1_2_VERSION:
+    case DTLS_ANY_VERSION:
+    case DTLS1_BAD_VER:
+        return 0;
+    default:
+        break;
+    }
+    ctx->quic_method = quic_method;
+    ctx->options &= SSL_OP_ENABLE_MIDDLEBOX_COMPAT;
+    return 1;
+}
+
+int SSL_set_quic_method(SSL *ssl, const SSL_QUIC_METHOD *quic_method)
+{
+    switch (ssl->method->version) {
+    case DTLS1_VERSION:
+    case DTLS1_2_VERSION:
+    case DTLS_ANY_VERSION:
+    case DTLS1_BAD_VER:
+        return 0;
+    default:
+        break;
+    }
+    ssl->quic_method = quic_method;
+    ssl->options &= SSL_OP_ENABLE_MIDDLEBOX_COMPAT;
+    return 1;
+}
+
+int quic_set_encryption_secrets(SSL *ssl, OSSL_ENCRYPTION_LEVEL level)
+{
+    uint8_t *read_secret = NULL;
+    uint8_t *write_secret = NULL;
+    static const unsigned char zeros[EVP_MAX_MD_SIZE];
+
+    if (!SSL_IS_QUIC(ssl))
+        return 1;
+
+    /* secrets from the POV of the client */
+    switch (level) {
+    case ssl_encryption_early_data:
+        write_secret = ssl->early_secret;
+        break;
+    case ssl_encryption_handshake:
+        read_secret = ssl->client_finished_secret;
+        write_secret = ssl->server_finished_secret;
```

Comment:
I'm not sure these secrets are correct.  As far as I tested, they are not handshake secrets.
There is another issue.  According to the documentation, set_encryption_secrets callback is supposed to be called with both read and write secret except for early data secret.  I'm writing in client perspective.  The problem is OpenSSL produces read secret (that is server_handshake_traffic) just after reading ServerHello, but it produces write secret (client_handshake_traffic) after reading 2nd server flight which is in Handshake packet in QUIC.
So at the moment, the first call of the callback provides write_secret with all zero, which is not usable.
I'm under impression that the callback is called just once per encryption level, so providing zero secret is not an option.  I think we need to change the code so that client_handshake_traffic secret is generated when ServerHello is read.  According to RFC8446, it is possible.

Reply:
This has been tested with Chromium QUIC, but I don't know which version. This may be the case, that only one secret is available at any one particular time. In such a case, you should wait for both to be available. On the other hand, the secrets could be checked to see if they are both non-zero before issuing the callback.

Reply:
Waiting for both secret does not work in OpenSSL.  In order to get handshake write_secret, you have to read CRYPTO data in handshake packet.  And in order to decrypt handshake packet, you need handshake read_secret.  Without moving the generation of write_secret at the same time when read_secret is available, waiting both gets dead lock.

Reply:
In BoringSSL, the callback is called just once per level and we derive both keys at the same time.

We did this so the caller doesn't have to deal with a half-initialized level. In particular, QUIC generally ACKs packets at the same encryption level they were sent at. If you have write keys without read keys, you will send packets that you cannot read ACKs to. That's weird. If you have read keys without write keys, you may receive a packet that you cannot write an ACK to. That's extremely weird and will likely cause the caller to crash in confusion.

Reply:
@davidben Understood. I believe we have this code working. I'm reaching out to our QUIC developer to see his current status. This PR may be a bit out of date.

Reply:
So we do check to make sure both secrets are available. Again, this has been tested with Chromium QUIC, but I believe as a server-size.

Reply:
> So at the moment, the first call of the callback provides write_secret with all zero, which is not usable.

The code at line 229 (https://github.com/openssl/openssl/pull/8797/files#diff-570136eab3b887f70e617f3b37451f3aR229) should be preventing this from happening. But @tatsuhiro-t you claim this is an issue due to the delay?

Reply:
I'm concerned that to generate both secrets at the same time will require a rewrite of `tls13_change_cipher_state()`. This will take a while (given that this is one of many tasks on my plate!). I welcome any input or suggestions.

Reply:
I think that after calling s->method->ssl3_enc->change_cipher_state(s, SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_READ) (in tls_process_server_hello), successively call s->method->ssl3_enc->change_cipher_state(s, SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE).  Remove any other invocation of change_cipher_state(s, SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE).  And then require both handshake secret nonzero in quic_set_encryption_secrets.
This at least makes QUIC handshake successful, but I'm not sure about the side effects.

I think OpenSSL developer knows better than me. cc @mattcaswell

Reply:
We have to move the generation of client_application_traffic earlier as well, otherwise server cannot write 1RTT packet before reading Client Finished.
This is a bit tricky because tls13_change_cipher_state currently produces client_application_traffic and resumption master secret in the same call.

Reply:
After some trial and errors, I found that it is relatively easier to produce just both read and write secrets in certain combination of "which" flags in tls13_change_cipher_state.  QUIC does not need key and iv at this point.  The changes are small and don't affect other TLSv1.3 path.

For example, produce client_handshake_traffic and server_handshake_traffic secret if label is server_handshake_traffic and SSL_CC_HANDSHAKE and (SSL3_CC_CLIENT and SSL_CC_READ or SSL_CC_SERVER and SSL_CC_WRITE).
For client_application_traffic and server_application_traffic secret if label is server_application_traffic and !SSL_CC_HANDSHAKE and (SSL3_CC_CLIENT and SSL_CC_READ or SSL_CC_SERVER and SSL_CC_WRITE).

And only call quic_set_encryption_secrets in these cases which eliminates zero compare.

Reply:
This has all been rewritten